include/boost/math/differentiation/autodiff.hpp

```diff
@@ -1786,9 +1784,9 @@ struct promote_args_2<differentiation::detail::fvar<RealType0, Order0>,
 {
   using type = differentiation::detail::fvar<typename promote_args_2<RealType0, RealType1>::type,
 #ifndef BOOST_NO_CXX14_CONSTEXPR
-           std::max(Order0,Order1)>;
+    (std::max)(Order0,Order1)>;
```

Comment:
Curious why the parentheses? `(std::max)` Does it fix a warning?

Reply:
MSVC requires it, otherwise you can pick up their std::min/std::max macros. Boost's developer guidelines touch on this (more specifically to `std::numeric_limits<>::max()`/`std::numeric_limits<>::min()`). The alternative is to define `NOMINMAX`, so `Windows.h` won't define any of these macros.
[Boost Developer Guidelines](https://github.com/boostorg/boost/wiki/Guidelines:-Warnings)

Reply:
If it's not required or satisfies a warning, I'd prefer to not bend the code to satisfy Windows.

If there is a warning, an alternative would be to try changing `BOOST_NO_CXX14_CONSTEXPR` to `BOOST_NO_CXX17_IF_CONSTEXPR` in case that helps (just a guess).

Reply:
Windows.h contains this macro ```#define max(a,b)            (((a) > (b)) ? (a) : (b))```
As such if a user were to have this code
```
#include <windows.h>
#include <autodiff.hpp>
```
then the expression ```std::max(Order0,Order1)``` resolves to ```std::(((Order0) > (Order1)) ? (Order0) : (Order1))```
in a sample program I whipped up, this gives the error ```Error	C2589	'(': illegal token on right side of '::'```

This is discussed here: https://stackoverflow.com/questions/11544073/how-do-i-deal-with-the-max-macro-in-windows-h-colliding-with-max-in-std . Kedarbhat's approach is the second answer.

The parentheses around max basically say "this isn't a macro". The macro definition in windows.h is truly awful but not something easily changed. Without the parentheses the options are
- Don't completely support windows
- Require your windows users to be careful about the relative ordering of these two headers
- #undef min/max in your header. This has the unfortunate consequence of affecting everything after the autodiff include
- Force windows users to use the NOMINMAX before they include windows. This may break their code and may be very difficult for them if they're not including windows.h directly (e.g. in the worst case if they're getting it via some third party header which depends on the macro).

Reply:
Thank you for the clear explanation of the the issue. It seems that from the Boost Developer Guidelines link Kedar provided, it recommends adding the `/DNOMINMAX` compiler switch which I very much agree with.

As you probably saw, I'd like to see what the boost guys say about this problem as well.

Reply:
That document is quite peculiarly structured but that's not the take away message I got from that section at all. Note this paragraph in particular:
> So it is far better to check that clashes are correctly avoided by bracketing as above because others may use your code when MIN/MAX macros are defined rather than using this compiler option.)

Nonetheless this document way well be dated so will be interesting to see what the boots guys say.